core/services/blockhashstore/feeder.go

```diff
@@ -0,0 +1,190 @@
+		toBlock          = int(latestBlock) - f.waitBlocks
+		blockToRequests  = make(map[uint64]map[string]struct{})
+		requestIDToBlock = make(map[string]uint64)
+	)
+	if fromBlock < 0 {
+		fromBlock = 0
+	}
+	if toBlock < 0 {
+		toBlock = 0
+	}
+	reqs, err := f.coordinator.Requests(ctx, uint64(fromBlock), uint64(toBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF requests",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return errors.Wrap(err, "fetching VRF requests")
+	}
+	for _, req := range reqs {
+		if _, ok := blockToRequests[req.Block]; !ok {
+			blockToRequests[req.Block] = make(map[string]struct{})
+		}
+		blockToRequests[req.Block][req.ID] = struct{}{}
+		requestIDToBlock[req.ID] = req.Block
+	}
+
+	fuls, err := f.coordinator.Fulfillments(ctx, uint64(fromBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF fulfillments",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return errors.Wrap(err, "fetching VRF fulfillments")
+	}
+	for _, ful := range fuls {
+		requestBlock, ok := requestIDToBlock[ful.ID]
+		if !ok {
+			continue
+		}
+		delete(blockToRequests[requestBlock], ful.ID)
+	}
+
+	var errs error
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) > 0 {
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				continue
+			}
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"error", err,
+					"block", block)
+				errs = multierr.Append(errs, errors.Wrap(err, "checking if stored"))
+			} else if stored {
+				f.logger.Infow("Blockhash already stored",
+					"block", block, "latestBlock", latestBlock,
+					"unfulfilledReqIDs", limitReqIDs(unfulfilledReqs))
+				f.stored[block] = struct{}{}
+				continue
+			}
+
+			// Block needs to be stored
+			err = f.bhs.Store(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to store block", "error", err, "block", block)
+				errs = multierr.Append(errs, errors.Wrap(err, "storing block"))
+				continue
+			}
+
+			f.logger.Infow("Stored blockhash",
+				"block", block, "latestBlock", latestBlock,
+				"unfulfilledReqIDs", limitReqIDs(unfulfilledReqs))
+			f.stored[block] = struct{}{}
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < uint64(fromBlock); block++ {
+			if _, ok := f.stored[block]; ok {
+				delete(f.stored, block)
+				f.logger.Debugw("Pruned block from stored cache",
+					"block", block, "latestBlock", latestBlock)
+			}
+		}
+	}
+	f.lastRunBlock = latestBlock
+	return errs
+}
+
+// limitReqIDs converts a set of request IDs to a slice limited to 50 IDs max.
+func limitReqIDs(reqs map[string]struct{}) []string {
```

Comment:
I'm not sure what the constraints are for logging but figured that once we're dealing with thousands of requests it could be an issue, I might be wrong though.

Reply:
ok, seems fine. I think the bptxm logs at least give us a way to determine whether we send Store's or not

Reply:
Yeah, and for stores we do have an info log for each one, it's just requests that are tricky cause there could be many in a single block